components/brave_wallet/browser/solana_provider_impl.cc

```diff
@@ -34,20 +45,42 @@ void SolanaProviderImpl::Init(
 
 void SolanaProviderImpl::Connect(absl::optional<base::Value> arg,
                                  ConnectCallback callback) {
-  // TODO(darkdh): handle onlyIfTrusted when it exists
   DCHECK(delegate_);
+  if (arg.has_value()) {
+    const base::Value::Dict* arg_dict = arg->GetIfDict();
+    if (arg_dict) {
+      absl::optional<bool> only_if_trusted =
+          arg_dict->FindBool(kOnlyIfTrustedOption);
+      if (only_if_trusted.has_value() && *only_if_trusted) {
+        delegate_->IsSelectedAccountAllowed(
+            mojom::CoinType::SOL,
+            base::BindOnce(&SolanaProviderImpl::OnEagerlyConnect,
+                           weak_factory_.GetWeakPtr(), std::move(callback)));
+        return;
+      }
+    }
+  }
   delegate_->RequestSolanaPermission(
       base::BindOnce(&SolanaProviderImpl::OnConnect, weak_factory_.GetWeakPtr(),
                      std::move(callback)));
 }
 
 void SolanaProviderImpl::Disconnect() {
-  NOTIMPLEMENTED();
+  DCHECK(keyring_service_);
+  absl::optional<std::string> account =
+      keyring_service_->GetSelectedAccount(mojom::CoinType::SOL);
+  if (account)
+    connected_map_.erase(*account);
 }
 
 void SolanaProviderImpl::IsConnected(IsConnectedCallback callback) {
-  NOTIMPLEMENTED();
-  std::move(callback).Run(false);
+  DCHECK(keyring_service_);
+  absl::optional<std::string> account =
+      keyring_service_->GetSelectedAccount(mojom::CoinType::SOL);
+  if (!account)
+    std::move(callback).Run(false);
+  else
+    std::move(callback).Run(connected_map_.contains(*account));
```

Comment:
I wonder maybe we should always get the actual permission status instead of having this local map?
AFAIK, `SolanaProviderImpl` is for per frame host, if I open two tabs to the same dapp site, I suppose the status of this connected_map_ would be different and will be inconsistent with the actual permission?
Also, what if user revokes the permission from settings UI? Wouldn't we wrongly still return connected here?
BTW, we need to have a test to cover when the permission is changed outside of this flow and see if APIs here still work as expected.

Reply:
no, the connected status is a non persistent status per frame for connect associated API to indicate the a site has call connect or not, it is absolutely separated from site permission.
For example, two tabs with same origin already have site permission, but one tab call connect and another one doesn't, then the one called connect will have connected status, and another one will show disconnect status. I will put this in the comment of connected map

Reply:
Okay, got it, please put detailed comments to describe how permission is used and this non persistent per frame connect status as their API is super confusing for people to understand, thanks.

Reply:
comment added in eb6ab5a49db53e721425ee6281fa8ab6a79e0a25